node/actors/bft/src/leader/replica_prepare.rs

```diff
@@ -132,34 +132,34 @@ impl StateMachine {
         // ----------- All checks finished. Now we process the message. --------------
 
         // We add the message to the incrementally-constructed QC.
-        self.prepare_qcs
+        let prepare_qc = self
+            .prepare_qcs
             .entry(message.view.number)
-            .or_insert_with(|| validator::PrepareQC::new(message.view.clone()))
-            .add(&signed_message, self.config.genesis());
+            .or_insert_with(|| validator::PrepareQC::new(message.view.clone()));
+        prepare_qc.add(&signed_message, self.config.genesis());
 
         // We store the message in our cache.
         self.prepare_message_cache
             .entry(message.view.number)
             .or_default()
-            .insert(author.clone(), signed_message);
-
-        // Now we check if we have enough messages to continue.
-        let num_messages = self
-            .prepare_message_cache
-            .get(&message.view.number)
-            .unwrap()
-            .len();
-
-        if num_messages < self.config.genesis().validators.threshold() {
+            .insert(author.clone(), signed_message.clone());
+
+        // Now we check if we have enough weight to continue.
+        let weight = self.config.genesis().validators.weight(
+            prepare_qc
+                .map
+                .get(&signed_message.msg)
+                .unwrap_or(&Signers::new(self.config.genesis().validators.len()))
+                .clone(),
+        );
```

Comment:
here you should sum weights over all messages collected for the given view, not just the one just received

Reply:
and please add a test for it

Reply:
mmhh, I believe that only the weights for the message received are able to exceed the threshold 🤔

Reply:
prepareqc consists of signatures on potentially different messages, which just share view number. It is the total weight over all such messages that is supposed to exceed threshold.

Reply:
+1

Reply:
Fixed, and added a test case for it